posts/regular-expressions-in-java.md

```diff
@@ -0,0 +1,177 @@
+* `PatternSyntaxException` — this class is used to define a syntax error within the regex pattern.
+* `MatchResult Interface` — this interface produces the results of the pattern matching. 
+
+Now that we know what package to import, we can start working to create our program. 
+
+First, we’ll need to import our packages, as mentioned earlier. For this tutorial, we’ll use “Scanner,” “Pattern,” and “Matcher.” From there, we will add some boilerplate to add our name, the date, and the title of our program. Then, we’ll initialize the `Main` class. 
+
+<PrismSyntaxHighlighter
+input={`package com.company;
+import java.util.Scanner;
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+ 
+// author: 
+// date: 
+// title: Java Regex for Email Validation
+ 
+public class Main {
+    public static void main(String[] args) {
+ 
+        System.out.println("Please enter a valid email address: ");
+        Scanner sc = new Scanner(System.in);
+        String email = sc.nextLine();
+        // now that we've told the program to save user input as an 
+        // "email" String object, we can add our regex pattern
+    }
+}`}
+language='java'
+/>
+
+In our `Main` class, we are providing output to guide the user to enter an email address. Then we are having Java take in user input from the next line, and assigning that input to the `email` string. 
+
+With this initialized, we will move onto setting up the regular expression to validate emails.
+
+## Adding `Pattern` and `Matcher`
+
+Next, we’ll need to enter our regex pattern with the `Pattern` class. The regex we’re using for email validation is the [RFC 5322 “General Email Regex” Official Standard](https://www.ietf.org/rfc/rfc5322.txt). We’ll copy it here, as it’s fairly long. You will have to horizontally scroll to get the full string.
+
+```
+^(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])$
+```
+
+This statement can be overwhelming if you don’t have much experience with regular expressions. You can review the “[Email Validation Summary](https://emailregex.com/email-validation-summary/)” from _Almost Perfect Email Regex_ to learn more. 
+
+Now, we’ll add this statement into our `Pattern` method. We’ll be building off the program we started above, working within the `Main` class.
+
+```
+...
+        String email = sc.nextLine();
+        Pattern pattern = Pattern.compile("^[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$", Pattern.CASE_INSENSITIVE);
```

Comment:
This is very funky because of the backslashes, so I put it into the regular markdown code block. This is not an ideal solution, was having trouble escaping the double backslashes in this string. Tried `\\\\`.

Reply:
Also tried `\'`